File: src/services/ranking/interp.ts
/**
 * Piecewise-linear interpolation over sorted anchor points, clamped at the ends.
 * Used to map a raw metric (e.g. a bodyweight-strength ratio) onto the 0..100
 * score scale via provisional anchor tables.
 */
export interface Anchor {
  x: number;
  y: number;
}

/** Interpolate y for a given x across anchors sorted ascending by x. */
export function interpolate(anchors: readonly Anchor[], x: number): number {
  if (anchors.length === 0) return 0;
  if (x <= anchors[0].x) return anchors[0].y;
  const last = anchors[anchors.length - 1];
  if (x >= last.x) return last.y;
  for (let i = 0; i < anchors.length - 1; i++) {
    const a = anchors[i];
    const b = anchors[i + 1];
    if (x >= a.x && x <= b.x) {
      const t = (x - a.x) / (b.x - a.x);
      return a.y + t * (b.y - a.y);
    }
  }
  return last.y;
}

File: android-native/app/src/main/java/com/fitnessrpg/app/domain/ranking/Interp.kt
package com.fitnessrpg.app.domain.ranking

/**
 * Piecewise-linear interpolation over sorted anchor points, clamped at the ends.
 * Maps a raw metric (e.g. a bodyweight-strength ratio) onto the 0..100 scale.
 */
data class Anchor(val x: Double, val y: Double)

/** Interpolate y for a given x across [anchors] sorted ascending by x. */
fun interpolate(anchors: List<Anchor>, x: Double): Double {
    if (anchors.isEmpty()) return 0.0
    if (x <= anchors.first().x) return anchors.first().y
    val last = anchors.last()
    if (x >= last.x) return last.y
    for (i in 0 until anchors.size - 1) {
        val a = anchors[i]
        val b = anchors[i + 1]
        if (x >= a.x && x <= b.x) {
            val t = (x - a.x) / (b.x - a.x)
            return a.y + t * (b.y - a.y)
        }
    }
    return last.y
}
